_posts/2022-08-22-fan-out-and-back-in-using-functions.md

```diff
@@ -0,0 +1,471 @@
+
+We're increasingly hearing that OpenFaaS functions are convenient for ingesting, transforming and processing large amounts of data. Functions are a natural fit for processing data because they're short-lived, stateless, efficient and can scale out to processing large amounts of data in parallel.
+
+A common pattern used when processing large amounts of data is to break down an expensive task into smaller sub tasks that can be executed in parallel. This fan-out pattern is already available in OpenFaaS through asynchronous functions. With the async feature, invocations are queued up and processed as and when there's capacity.
+
+> In a previous article, we showed you how to process large amounts of data both asynchronously and reliably with the fan-out pattern: [How to process your data the resilient way with back pressure ](https://www.openfaas.com/blog/limits-and-backpressure/).
+
+The MapReduce pattern was popularised by Hadoop, which specialises in processing large batches of data. One dataset or task is split into many others, which can be worked on in parallel over a large distributed network of computers, that's the map step. The reduce step is where the results are combined to make sense of them.
+
+We've seen customers making use of OpenFaaS in this way, so we wanted to show you how we think it can be done without making any changes to OpenFaaS itself.
+
+## The fan-out pattern
+Fan-out can be used to split a larger task or batch of tasks into smaller sub tasks. The processing of each sub task is deferred to another function. These functions can run in parallel to complete all tasks as quickly and efficiently as possible.
+
+> The fan-out pattern is already supported in OpenFaaS through the [asynchronous invocation system](https://docs.openfaas.com/reference/async/)
+
+![fan-out conceptual diagram](/images/2022-fan-out-and-back-in-using-functions/fan-out.png)
+> Diagram of fan-out with functions
+
+## The fan-in pattern
+The fan-in pattern can be applied if you need to wait for all sub tasks to complete before moving on to the next processing step. It can be used to collect and combine the result from each individual sub task.
+
+To fan-in, some kind of shared storage, like a database, a Redis key or an S3 bucket is required for tracking progress of the individual sub tasks.
+
+![fan-in conceptual diagram](/images/2022-fan-out-and-back-in-using-functions/fan-in.png)
+> Diagram of fan-in with functions
+
+## Implement a fan-out/fan-in pattern in OpenFaaS.
+In this section we are going to show you how this pattern can be implemented with OpenFaaS through a relatable use-case. A common big data scenario is batch processing of a data set. In this scenario data is retrieved from a storage system and then processed by parallelized jobs. The results of each individual job is persisted in some kind of data store where it can later be retrieved for further processing when the batch is completed.
```

Comment:
`The results of each individual job is persisted` becomes either:
`The result of each individual job is persisted` or 
`The results of each individual job are persisted`

Reply:
Thanks for the suggestions. I used the first one.